File: course_level1_py/course_level1_py/number_counter.py
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool

class RobotNewsStationNode(Node):
    def __init__(self):
        super().__init__("number_counter")
        self.sub_ = self.create_subscription(Int64, "number", self.callback_number, 10)
        self.pub_ = self.create_publisher(Int64, "number_count", 10)
        self.create_service(SetBool, "reset_counter", self.callback_reset_counter)
        self.counter_ = 0
        self.get_logger().info("Number Counter has been started!")

    def callback_number(self, msg: Int64):
        self.counter_ += msg.data
        new_msg = Int64()
        new_msg.data = self.counter_
        self.pub_.publish(new_msg)

    def callback_reset_counter(self, request: SetBool.Request, response: SetBool.Response):
        if (request.data): 
            self.counter_ = 0
            response.success = True
            response.message = "Counter reset to 0"
            return response
        response.success = False
        response.message = "Counter not reset to 0"
        return response

def main(args=None):
    rclpy.init(args=args)
    node = RobotNewsStationNode()
    rclpy.spin(node)
    rclpy.shutdown()
    
    
if __name__ == "__main__":
    main()

File: course_level1/course_level1_cpp/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

using namespace std::placeholders;
    
class NumberCounterNode : public rclcpp::Node
{
public:
    NumberCounterNode() : Node("number_counter"), number_counter_(0)
    {
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
            "number", 10, 
            std::bind(&NumberCounterNode::callbackNumberSub, this, _1));
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>(
            "number_count", 10);
        service_ = this->create_service<example_interfaces::srv::SetBool>(
                "reset_counter", std::bind(&NumberCounterNode::callbackResetCounter, this, _1, _2));
    }
     
private:
    void callbackNumberSub(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        number_counter_ += msg->data;
        auto msg2 = example_interfaces::msg::Int64();
        msg2.data = number_counter_;
        publisher_->publish(msg2);
    }

    void callbackResetCounter(const example_interfaces::srv::SetBool::Request::SharedPtr request, 
                         const example_interfaces::srv::SetBool::Response::SharedPtr response)
    {
        if (request->data) {
            number_counter_ = 0;
            response->success = true;
            response->message = "Counter reset to 0";
        } else {
            response->success = false;
            response->message = "Counter not reset to 0";
        }
    }

    int number_counter_;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr service_;
};
    
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}